Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm

```diff
@@ -127,8 +159,44 @@ static double magnitudeToIntensity(double magnitude)
         RELEASE_LOG_ERROR(Gamepad, "GameControllerHapticEffect::start: Failed to start the weak player");
         m_rightPlayer = nullptr;
     }
-    if (!m_leftPlayer && !m_rightPlayer)
+    if (!m_leftPlayer && !m_rightPlayer) {
         m_completionHandler(false);
+        return;
+    }
+
+    if (m_leftPlayer) {
+#if USE(HAPTICS_COMPLETION_HANDLER)
+        [m_leftEngine.get() notifyWhenPlayersFinished:makeBlockPtr([weakThis = WeakPtr { *this }](NSError *) mutable -> CHHapticEngineFinishedAction {
+            ensureOnMainRunLoop([weakThis = WTFMove(weakThis)] {
+                if (weakThis)
+                    weakThis->leftEffectFinishedPlaying();
+            });
+            return CHHapticEngineFinishedActionStopEngine;
+        }).get()];
+#else
+        RunLoop::main().dispatchAfter(m_delay + m_duration, [weakThis = WeakPtr { *this }] {
+            if (weakThis)
+                weakThis->leftEffectFinishedPlaying();
+        });
+#endif
+    }
+
+    if (m_rightPlayer) {
+#if USE(HAPTICS_COMPLETION_HANDLER)
+        [m_rightEngine.get() notifyWhenPlayersFinished:makeBlockPtr([weakThis = WeakPtr { *this }](NSError *) mutable -> CHHapticEngineFinishedAction {
+            ensureOnMainRunLoop([weakThis = WTFMove(weakThis)] {
+                if (weakThis)
+                    weakThis->rightEffectFinishedPlaying();
+            });
+            return CHHapticEngineFinishedActionStopEngine;
+        }).get()];
+#else
+        RunLoop::main().dispatchAfter(m_delay + m_duration, [weakThis = WeakPtr { *this }] {
```

Comment:
This doesn't seem like something we should ship. It looks like a workaround for a bug we haven't figured out yet.